src/lib/statistics/chunk_statistics/histograms/abstract_histogram.cpp

```diff
@@ -0,0 +1,620 @@
+       * In practice, _bin_count(n) >= _count_per_bin for n < bin_count() - 1,
+       * because bins are filled up until the count is at least _count_per_bin.
+       * The last bin typically has a count lower than _count_per_bin.
+       * Therefore, if we calculate the share of the last bin based on _count_per_bin
+       * we might end up with an estimate higher than total_count(), which is then capped.
+       */
+      return std::min(cardinality, static_cast<float>(total_count()));
+    }
+    case PredicateCondition::LessThanEquals:
+      return estimate_cardinality(PredicateCondition::LessThan, _get_next_value(value));
+    case PredicateCondition::GreaterThanEquals:
+      return total_count() - estimate_cardinality(PredicateCondition::LessThan, variant_value);
+    case PredicateCondition::GreaterThan:
+      return total_count() - estimate_cardinality(PredicateCondition::LessThanEquals, variant_value);
+    case PredicateCondition::Between: {
+      Assert(static_cast<bool>(variant_value2), "Between operator needs two values.");
+      const auto value2 = type_cast<T>(*variant_value2);
+
+      if (value2 < value) {
+        return 0.f;
+      }
+
+      return estimate_cardinality(PredicateCondition::LessThanEquals, *variant_value2) -
+             estimate_cardinality(PredicateCondition::LessThan, variant_value);
+    }
+    case PredicateCondition::Like:
+    case PredicateCondition::NotLike:
+      Fail("Predicate NOT LIKE is not supported for non-string columns.");
+    default:
+      // TODO(anyone): implement more meaningful things here
+      return total_count();
+  }
+}
+
+// Specialization for numbers.
+template <typename T>
+float AbstractHistogram<T>::estimate_cardinality(const PredicateCondition predicate_type,
+                                                 const AllTypeVariant& variant_value,
+                                                 const std::optional<AllTypeVariant>& variant_value2) const {
+  return _estimate_cardinality(predicate_type, variant_value, variant_value2);
+}
+
+// Specialization for strings.
+template <>
+float AbstractHistogram<std::string>::estimate_cardinality(const PredicateCondition predicate_type,
+                                                           const AllTypeVariant& variant_value,
+                                                           const std::optional<AllTypeVariant>& variant_value2) const {
+  const auto value = type_cast<std::string>(variant_value);
+
+  // Only allow supported characters in search value.
+  // If predicate is (NOT) LIKE additionally allow wildcards.
+  const auto allowed_characters =
+      _supported_characters +
+      (predicate_type == PredicateCondition::Like || predicate_type == PredicateCondition::NotLike ? "_%" : "");
+  Assert(value.find_first_not_of(allowed_characters) == std::string::npos, "Unsupported characters.");
+
+  if (can_prune(predicate_type, variant_value, variant_value2)) {
+    return 0.f;
```

Comment:
Does an estimated cardinality of 0 always mean "prunable"? If so, can't we combine the two places where LIKEs are handled?

Reply:
Not sure I follow.
Theoretically, because of float rounding, it could happen that `can_prune(x) == false && estimate_cardinality(x) == 0.f`.
Which two places to be combined do you mean?

Reply:
That was my question, if `can_prune == true <==> estimate_cardinality == 0`. If the latter could convey "guaranteed 0", we wouldn't need the `can_prune` method, correct?

Just an idea (not convinced myself yet): What if negative zero means "guaranteed 0"?

Reply:
Yes, we _could_ combine them if estimate_cardinality could convey a guarantee.
This could, by the way, also be interesting for "we know that *all* tuples match the predicate", and possibly just forward a chunk in some operators or whatever.
From a quick google search it looks like `-0.0f == 0.0f`, so how would you differentiate them?
One possibility would be to return `std::pair<bool, float>`, where `std::pair<true, 0.f>` means can be pruned, `std::pair<true, 1.f>` means "all values match", and `std::pair<false, x>` means x is an estimated selectivity (or cardinality).

Reply:
I like the pair idea. Maybe `std::pair<float, bool>` where the second value means "I am really sure about this"?

Reply:
Good idea, because histograms can actually sometimes be 100% sure even about selectivities other than 0 and 1.
I would implement that as part of the new statistics component, though, because it has implications on other parts as well. Are you fine with that?

Reply:
Ok, is this tracked somewhere?

Reply:
#1153